src/cr/cube/dimension.py

```diff
@@ -976,3 +998,120 @@ def label(self):
     def prune(self):
         """True if this subtotal should not appear when empty."""
         return self._prune
+
+
+class _BaseSmoother(object):
+    """Base class for smoothing objects"""
+
+    def __init__(self, dimension_transforms_dict, is_cat_date):
+        self._dimension_transforms_dict = dimension_transforms_dict
+        self._is_cat_date = is_cat_date
+
+    @classmethod
+    def factory(cls, dimension_transforms_dict, is_cat_date):
+        """Returns appropriate Smoother object according to transforms_dict"""
+        # --- smoothing is only applied to CAT_DATE dimensions ---
+        if not is_cat_date:
+            return _NullSmoother()
+
+        # --- smoothing is only applied when requested ---
+        if not cls._show_smoothing(dimension_transforms_dict):
+            return _NullSmoother()
+
+        # --- otherwise smoothing is single-side moving-average ---
+        return _SingleSideMovingAvgSmoother(dimension_transforms_dict)
```

Comment:
On third thought, this can just look like this. The key points are:

* The entire responsibility of the (single) smoothing class is to form a _smoothing function_ configured with the right window (when applicable).

* `Dimension` has a single public property `.smooth` which is ("returns") a _function_, so the client calls `dimension.smooth(values)` and does not need to get or manage a smoother instance.

```python
class Dimension(...):
   ...
   @lazyproperty
   def smooth(self):
       """Function performing smoothing for this dimension, based on transform."""

       def null_smooth(values):
           return values

       if not self._show_smoothing:
           return null_smooth

       return _MovingAvgSmoother.smoothing_function(self._smoothing_window)


class _MovingAvgSmoother(object):
    """Creates and configures a smoothing function."""

    def __init__(self, window):
        self._window = window

    @classmethod
    def smoothing_function(cls, window):
        """Returns function that smooths a sequence of numeric values in `window`.

        ... more specifics about calling signature of returned function and what
        window means and np.nan values at beginning, etc.
        """
        return cls(window)._smoothing_function

    @lazyproperty
    def _smoothing_function(self):
        """smoothing function configured with window ... """

        def smooth(values):
            """function returned to ..."""
            smoothed_values = self._smoother(values)
            ...
            return np.concatenate(...)

        return smooth

    def _smoother(self, values):
        """ ... """
        window = self._window
        return (
            np.array(tuple(np.convolve(...)
            if ...
            else np.array(
                ...
            )
        )
```